File: java how to program-PersonalSolution/case study/7.18/GradeBook.java
public class GradeBook

  {

         private String courseName;
         private int [][] grades;

        
         public GradeBook(String name,int [][] gradesArray)
           {
              courseName = name;
              grade = gradesArray;

          }

         
        public void setCourseName(String name)
          {
              courseName = name;
          }

        public get getCourseName()
          {
              return courseName;
          }

        public void displayMessage()
        {
           
            System.out.printf("Welcome to the grade book for \n%s!\n\n",getCourseName() );
        }

        public void processGrades()//------------------------------------------------------------------------------------------------------------------------
        {
              outputGrades();

              System.out.printf("\n%s %d\n%s %d\n\n","Lowest grade in the grade book is", getMinimum(),"Highest grade in the grade book is",getMaximum() );
        
              outputBarChart()
         }//--------------------------------------------------------------------------------------------------------------------------------------------

        public int getMinimum()
         {
            int lowGrade = grades[0][0];

          for(int [] studentGrades : grades)
         {
            for(int grade : studentsGrades)
             {
                if(grade < lowGrade)
                  lowGrade = grade;
             }
           }
           return lowGrade;
         }

         public int getMaximum()
         {
           
            highGrade = grades[0][0];
          for(int [] studentsGrades : grades)
         {
            for(int grade : studentsGrades)
              {
                  if(grade > highGrade)
                    {
                        highGrade = grade;
                     }
               }
          }
           return highGrade;
         }

          public double getAverage(int [] setOfGrades)
             {
                  int total = 0;

                for(int grade : setOfGrades)
                  {
                      total+= grade;
                  }

                 return total / setOfGrades.length;
              }

           public void outputBarChart()
             {
                System.out.println("Overall grade distribution: ");

                int [] frequency = new int [11];

                for(int [] studentsGrades : grades)
                 {
                    for(int grade : studentGrades)
                      {
                         ++frequency[grade/10];
                      }
                 }

                for ( int count = 0; count < frequency.length; count++)
                  {
                      if(count == 10)
                       System.out.printf("%5d: ",100);
                      else
                       System.out.printf("%02d-%02d: ",count *10,count*10 + 9);
                   

                        for(int stars = 0;stars < frequency[count];stars++)
                          {
                              System.out.print("*");
                          }

                       System.out.println();

                   }//outer for

             }//outputBarChart

             
             public void outputGrades()
              {

                   System.out.println("The grades are:\n");
                   System.out.print("           ");

                   for(int test = 0; test < grades[0].length; test++)
                     System.out.printf("Test %d ",test+1);

                   System.out.println("Average");

                   for(int student = 0; student < grades.length; student++)
                    {
                       System.out.printf("Student %2d", student + 1);

                       for(int test : grades[student] )
                         System.out.printf("%8d",test);
                        
                       double average = getAverage(grades[student]);
                       Sysem.out.printf("%9.2f\n",average);
                    }
               }
         }//class end
                       

File: DSA/arrays/c++HowToProgram/CaseStudy/GradeBook.cpp
#include <iostream>
#include <iomanip>
#include "GradeBook.h"
using namespace std;


GradeBook::GradeBook(string name, const int gradesArray[])
 {
	setCourseName(name);

	for(int grade=0;grade < students;++grade)
	{
		grades[grade] = gradesArray[grade];
	}
}

void GradeBook::setCourseName(string name)
{
	courseName = name;
}

string GradeBook::getCourseName()
{
	return courseName;
}

void GradeBook::displayMessage()
{
	cout << "Welcome to the grade book for\n" << getCourseName() << "!" << endl;
}

void GradeBook::processGrades()
{
	outputGrades();
	
	cout << "\nClass average is " << setprecision(2) << fixed << getAverage() << "\nLowest grade is " << getMinimum() << "\nHighest grade is " << getMaximum() << endl;

	outputBarChart();
}

int GradeBook::getMinimum()
{
		int lowGrade = 100;

		for(int grade = 0; grade < students;++grade)
		{
			if(grades[grade] < lowGrade)
						lowGrade = grades[grade];
		}

	return lowGrade
}

int GradeBook::getMaximum()
{
		int highGrade = 0;
	
		for(int grade = 0; grade < students; ++grade)
		{
			if(grades[grade] > highGrade)
						highGrade = grades[grade];
		}

	return highGrade;
}

double GradeBook::getAverage()
{
		int total = 0;

		for(int grade = 0; grade < students; ++grade)
		{
			total += grades[grade];
		}
	return static_cast <double> (total)/students;
}

void GradeBook::outputBarChart()
{
	cout << "\nGrade distribution:" << endl;

	const int frequencySize = 11;
	int frequency[frequencySize] = {};

	
	for(int grade = 0; grade < students; ++grade)
	{
		++frequency[grades[grade]/students];

		for(int count = 0; count < frequencySize ; ++count)
			{
				if(count == 0)
						cout << " 0-9: ";
				else if(count == 10)
						cout << " 100: ";
				else
					cout << count * 10 << "-" << (count *10) + 9 << ": ";

					for(int stars = 0; stars < frequency[count]; ++stars)
					{
						cout << '*';
					}
				cout << endl;
			}	
	}
}

void GradeBook::outputGrades()
{
	cout << "\nThe grades are:\n\n";

	for(int student = 0; student < students; ++student)
	{
		cout << "Student " << setw(2) << student + 1 << ": " << setw(3) << grades[student] << endl;
}

		
	
